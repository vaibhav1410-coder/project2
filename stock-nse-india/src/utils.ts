import * as Moment from 'moment';
import { extendMoment } from 'moment-range';

const moment = extendMoment(Moment);
/**
 * @private
 */
export const getDateRangeChunks = (startDate: Date, endDate: Date, chunkInDays: number) => {
    const range = moment.range(startDate, endDate)
    const chunks = Array.from(range.by('days', { step: chunkInDays }))
    const dateRanges = []
    for (let i = 0; i < chunks.length; i++) {
        dateRanges.push({
            start: i > 0 ? chunks[i].add(1, 'day').format('DD-MM-YYYY') : chunks[i].format('DD-MM-YYYY'),
            end: chunks[i + 1] ? chunks[i + 1].format('DD-MM-YYYY') : range.end.format('DD-MM-YYYY')
        })
    }
    return dateRanges
}
/**
 * @private
 */
export const sleep = (ms: number) => {
    return new Promise(resolve => {
        setTimeout(() => {
            resolve('')
        }, ms)
    })
}

/**
 * @private
 * @param obj 
 * @returns 
 */
export const getDataSchema = (data: any, isTypeStrict=true): any[] | string => {
    if (typeof data !== 'object')
        return isTypeStrict ? `${typeof data}` : 'any'
    
    if(Array.isArray(data) && typeof data[0] !== 'object'){
        return isTypeStrict ? `${typeof data[0]}[]` : 'any'
    }

    return Object.entries(data).map(([key, value]) => {
        if (Moment.isDate(value))
            return `${key}: ${isTypeStrict ? 'Date': 'any'}`

        if (value === null || typeof value === 'string')
            return `${key}: ${isTypeStrict ? 'string | null': 'any'}`

        if (typeof value !== 'string' && Array.isArray(value)) {
            const typeForEmpty = isTypeStrict ? [] : 'any'
            return {
                [`${key}`]: value.length ? getDataSchema(value[0],isTypeStrict) : typeForEmpty
            }
        }

        if (typeof value === 'object') {
            return {
                [`${key}`]: getDataSchema(value,isTypeStrict)
            }
        }

        return `${key}: ${isTypeStrict ? typeof value : 'any'}`
    })
}
