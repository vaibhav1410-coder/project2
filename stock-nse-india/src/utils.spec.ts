import { getDataSchema } from './utils'
describe('Utils', () => {
    test('getDataSchema with type strict',()=>{
        const data={
            a:1,
            b: 'b',
            c: [1,2,3],
            d: {a:1,b:[2,3,4],c:{d:1}},
            e: new Date('2022-09-12')
        }
        expect(getDataSchema(data)).toMatchSnapshot()
        expect(getDataSchema(['hi'])).toMatchSnapshot('array_validation')
    })
    test('getDataSchema for date value having typeStrict:false',()=>{
        const data={
            date: new Date('2022-09-12')
        }
        expect(getDataSchema(data,false)).toMatchSnapshot()
    })
})