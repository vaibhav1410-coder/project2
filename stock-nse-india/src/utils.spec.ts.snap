// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Utils getDataSchema for date value having typeStrict:false 1`] = `
Array [
  "date: any",
]
`;

exports[`Utils getDataSchema with type strict 1`] = `
Array [
  "a: number",
  "b: string | null",
  Object {
    "c": "number",
  },
  Object {
    "d": Array [
      "a: number",
      Object {
        "b": "number",
      },
      Object {
        "c": Array [
          "d: number",
        ],
      },
    ],
  },
  "e: Date",
]
`;

exports[`Utils getDataSchema with type strict: array_validation 1`] = `"string[]"`;
